executor/src/witgen/machines/block_machine.rs

```diff
@@ -295,8 +312,28 @@ impl<T: FieldElement> BlockMachine<T> {
 
         // Record the steps where we made progress, so we can report this to the
         // cache later on.
+        let mut progress_in_last_step = false;
         let mut progress_steps = vec![];
-        for step in sequence {
+
+        // We might or might not have a cached sequence. If not, we use the default one, which
```

Comment:
Isn't it the job of the sequence cache to give us a default sequence in case it does not have anything cached? We just get some iterator, we report back what worked and hope that the sequence cache can make use of that information.